### ISSUE TITLE
Atomtypes with the sames name but different parameters are not preserved during merge

### ISSUE THREAD

**Issue by msuruzhon** [michellab/BioSimSpace#300]
Hi,

I have discovered an issue, where if one tries to merge two molecules which have common atomtype names but different associated parameters, the parameters of the second molecule effectively get lost during writing out.

For example, let the first atomtype be as follows (as taken from a GROMACS top file):
```
H           1    1.007940    0.000000       A    0.500000    0.400000
```

and the second atomtype be as follows:
```    
H           1    1.007940    0.000000       A    0.300000    0.600000
```

The resulting merged topology only has the following definitions:
```
     H           1    1.007940    0.000000       A    0.500000    0.400000
  H_du           0    0.000000    0.000000       A    0.000000    0.000000
```

Files to reproduce this behaviour can be found [here](https://github.com/michellab/BioSimSpace/files/8630571/Archive.zip).

This is a problem that shouldn't occur when one uses GAFF, because of the strict atom type names. However, OpenFF doesn't have canonical atom type names, so an `H1` in one molecule will not in general be the same as an `H1` in another molecule. I think we need to figure out how to uniquify the atom types of the two molecules if they have different parameters but the same name.

**Comment by lohedges** [comment]
Thanks for this. Ironically I've just stumbled across this issue myself. In addition, there appear to be a few other wrinkles when using OpenFF, e.g. molecules being returned with no ligand name when using SMILES, which would break writing of GROMACS topology files. (See [here](https://github.com/michellab/BioSimSpace/issues/299) for details.

As you say, we'll need some tricks to uniquify the atom types to ensure that all atom definitions appear in the topology. I'll look into this when I get the chance. It is probably easiest to add this as a hard requirement of the merging code itself, rather than tweaking things when topology files are written. I guess we might need to be careful if specify atom names have special meanings for an engine, but this is generally been okay in practice. (Apart from naming of water molecules and dummy atoms, in my experience.)

Cheers.

**Comment by lohedges** [comment]
Having looked at the `Sire.IO.GroTop` parser I think there actually might be a bug in the way that the `[ atomtypes ]` record is created in that it only considers the types from the lambda=0 state. (This is irrespective of types with the same name having different parameters.) This won't be a problem when types in the lambda = 1 state also exist at lambda = 0, but would if when the lambda = 1 state contains unique types. I'll check this in more detail and update when I've figured things out. (We probably need to combine atomtypes in both end states to write the record.)

**Comment by lohedges** [comment]
No, I was wrong, we do inspect both end states. This is just an issue of dealing with atom types with the same name, but different parameters, which should be possible to solve in the `merge` function.

**Comment by lohedges** [comment]
I have a strategy to uniquify the `atomtype` properties, which should hopefully work for GROMACS. Just to check, does this cause issues with the AMBER FEP setup too? If so, I might need to think about uniquifying the `ambertype` property too.

**Comment by chryswoods** [comment]
I think that ambertype should be unique, as there is more agreement what those types mean (they are defined in the gaff or FFXX forcefields, and have unique names). Gromacs treats atom types differently, if I remember correctly.

**Comment by msuruzhon** [comment]
So far I can't reproduce this issue using only AMBER so we might be good for now. I will let you know if I find anything interesting.

**Comment by lohedges** [comment]
Great, thanks for confirming. I'll push a fix once I've had a chance to test a little further. At the moment it just adds an `x` to the `atomtype` property if a duplicate is found. I only check for duplicates twice, since (presumably) there would only be one duplicate of a given type, i.e. because we are merging two molecules. (It shouldn't be possible to have a duplicate within an end state.)

**Comment by msuruzhon** [comment]
> (It shouldn't be possible to have a duplicate within an end state.)

Ah, but that's another potential issue - OpenFF atom type names clash in general with protein atom type names so if both of them are combined, this will be the case. Or are these automatically renamed when two systems are combined?

**Comment by chryswoods** [comment]
I agree it is a can of worms. The ``AmberPrm`` parser assumes that the names refer to unique atom types (``%FLAG AMBER_ATOM_TYPE``). However, the amber types are not actually used for anything in a prmtop file, so this doesn't matter (they would only be used by, e.g. antechamber, to assign LJ parameters and bond/angle/dihedral parameters). The actual LJ parameters are used to get unique LJ types (``%FLAG ATOM_TYPE_INDEX``) which index into the arithmetically combined parameters in ``%FLAG NONBONDED_PARM_INDEX``. Similarly, the actual bond/angle/dihedral parameters are used to set the values in the prmtop file.

So, for ``AmberPrm`` it is an annoyance (different amber atom types will have different LJ parameters, which may be unexpected) but it is not breaking. For ``GroTop``,  it is an issue because this format requires that all atom types are consistent (have the same mass, vdw parameters, element etc). I think you are right that this needs fixing in the ``GroTop`` parser, which should detect if an atom type is not consistent, and then rename it as needed.

**Comment by chryswoods** [comment]
Painful. That will be really slow to do at the ``System`` level, because it is O(n^2) - every atom type in the added molecules will need checking against every atom type in the system. Adding molecules one by one, and performing the check each time, will be expensive.

**Comment by msuruzhon** [comment]
I wrote a similar function that fixes that in ParmEd and as long as you use sets, it is pretty fast even on protein-ligand systems, so I guess this shouldn't be that bad for Sire either? Usually there are not that many atom types anyway.

**Comment by chryswoods** [comment]
Yes, I agree it will be quick if we use sets, and only do this operation when we want to write the molecule to a file. Thanks :-)

**Comment by lohedges** [comment]
Actually, I think we might be okay with my original solution of simply fixing the `[ atomtypes ]` record in the topology file. I forgot that the `Gro87` file just contains a list of named atoms by index, which are matched against the `[ atoms ]` section from the `GroTop` file in order. The entries in the `Gro87` file give the atom names, not types, and it appears that it's fine to have duplicate names for different types. (At least `grommp` doesn't complain and the simulation runs without error.)

I'll try to think of a simple test to check that things are working as expected.
